packages/runtime/native/CSS.ml

```diff
@@ -300,68 +302,81 @@ let resolve_selectors rules =
-
-  let declarations =
-    match declarations with
-    | [||] -> ""
-    | _ ->
-      Printf.sprintf ".%s { %s }" className (render_declarations declarations)
-  in
-
-  let selectors =
-    match selectors with
-    | [||] -> ""
-    | _ ->
-      selectors
-      |> Array.filter_map ~f:(render_selectors className)
-      |> Array.join ~sep:" "
-  in
-
-  (* Trimming is necessary to ensure there isn't an empty space when one of `declarations` or `selectors` is empty. *)
-  String.trim @@ String.concat " " [ declarations; selectors ]
+  (match declarations with
+  | [||] -> ()
+  | _ ->
+    Buffer.add_char out '.';
+    Buffer.add_string out className;
+    Buffer.add_string out " { ";
+    render_declarations out declarations;
+    Buffer.add_string out " }");
+  match selectors with
+  | [||] -> ()
+  | _ ->
+    if Array.length declarations > 0 then Buffer.add_char out ' ';
+    selectors
+    |> Array.iter ~f:(fun selector -> render_selectors out className selector);
+    Buffer.truncate out (Buffer.length out - 1)
 
 (* Renders all selectors with the hash given *)
-and render_selectors hash rule =
+and render_selectors out hash rule =
   match rule with
-  | Rule.Selector (_selector, rules) when Array.is_empty rules -> None
+  | Rule.Selector (_selector, rules) when Array.is_empty rules -> ()
   (* In case of being @media (or any at_rule) render the selector first and declarations with the hash inside *)
   | Rule.Selector (selector, rules) when contains_at selector ->
-    let nested_selectors = render_rules hash rules in
-    Some (Printf.sprintf "%s { %s }" selector nested_selectors)
+    Buffer.add_string out selector;
+    Buffer.add_string out " { ";
+    render_rules out hash rules;
+    Buffer.add_string out " }";
+    Buffer.add_char out ' '
   | Rule.Selector (selector, rules) ->
-    let new_selector = resolve_ampersand hash selector in
-    (* Resolving the ampersand means to replace all ampersands by the hash *)
-    Some (Printf.sprintf "%s { %s }" new_selector (render_declarations rules))
+    Buffer.add_string out (resolve_ampersand hash selector);
+    Buffer.add_string out " { ";
+    render_declarations out rules;
+    Buffer.add_string out " }";
+    Buffer.add_char out ' '
   (* Declarations aren't there *)
-  | _ -> None
+  | _ -> ()
 
 (* rules_to_string renders the rule in a format where the hash matches with `@emotion/serialise`. It doesn't render any whitespace. (compared to render_rules) *)
 (* TODO: Ensure Selector is serialised correctly *)
-let rec rules_to_string rules =
-  let buff = Buffer.create 16 in
-  let push = Buffer.add_string buff in
-  let rule_to_string rule =
-    match rule with
-    | Rule.Declaration (property, value) ->
-      push (Printf.sprintf "%s:%s;" property value)
-    | Rule.Selector (selector, rules) ->
-      let rules = rules_to_string rules in
-      push (Printf.sprintf "%s{%s}" selector rules)
+let rules_to_string rules =
+  let out = Buffer.create 1024 in
```

Comment:
Just an idea, instead of passing the buffer everywhere, what do you think about passing a `~push` callback? Not suggesting to do the change, but rather to discuss if you see it better or worst.

Reply:
This could work, I guess the upside would be we are not dependent on Buffer.
Since we can push string and char, a possible solution would be using two callbacks `~pushc` `~pushs`, or maybe a polymorphic variant prop `` `string of string | `char of char `` ?

Reply:
meh, true. Not a big deal